File: 0039-combination-sum/0039-combination-sum.java




class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> curNode =  new ArrayList<Integer>();
        backtrack(0, candidates, target, curNode, res);
        return res;
    }

    private void backtrack(int s, int[] candidates, int target, List<Integer> curNode,
                   List<List<Integer>> ans) {
    if (target < 0)
      return;
    if (target == 0) {
      ans.add(new ArrayList<>(curNode));
      return;
    }

    for (int i = s; i < candidates.length; ++i) {
      curNode.add(candidates[i]);
      backtrack(i, candidates, target - candidates[i], curNode, ans);
      curNode.remove(curNode.size() - 1);
    }
  }
}

File: 0039-combination-sum/0039-combination-sum.cpp
/*
Input: candidates = [2,3,6,7], target = 7
            2,3,6,7
       
                    /.                                     \.       \       \
                   2.                       (               3.         6.   (7)
            /.          \.       \                           /   \.    / \
           2                3.     6,7 (not solution).      3.    5.   6  7
     /  /  /. \.         /.  \                             /\.     
     7  6 2  3 = (7).   3,6,7.     (not solution).        3,5,6,7  
       /. \ \ \.      
    2.    3. 6. 7
=8 >7     9>7 12 13 

*/

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res ;
        vector<int> curNode ;
        backtrack(0, candidates, target, curNode, res);
        return res;
    }

private:
    void backtrack(int s, vector<int>& candidates,  int target, vector<int>& curNode, vector<vector<int>>& res) {
        if (target < 0) {
            return;
        }
        if (target == 0 ){
            res.push_back(curNode);
            return;
        }

        for (int i = s; i < candidates.size(); ++i){
            curNode.push_back(candidates[i]);
            backtrack(i, candidates, target - candidates[i], curNode, res);
            curNode.pop_back();
        }
    }
};